Remove the mutation-based restriction checks

psarna opened the pull request:
This series unifies the interface for checking if CQL restrictions are satisfied. Previously, an additional mutation-based approach was added in the materialized views layer, but the decision was reached that it's better to have a single API based on partition slices. With that, the regular selection path gets simplified at the cost of more complicated view generation path, which is a good tradeoff.
Note that in order to unify the interface, the view layer performs ugly transformations in order to adjust the input for `is_satisfied_by`. Reviewers, please take a close look at this code (`matches_view_filter`, `clustering_prefix_matches`, `partition_key_matches`), because it looks error-prone and relies on dirty internals of our serialization layer. If somebody has a better suggestion on how to do the transformation, I'm all ears.

Tests: unit(release), manual(playing with materialized views with custom filters)
Fixes #7215

dekimir commented on db/view/view.cc:
```diff
@@ -259,15 +261,34 @@ void stats::register_stats() {
 
 bool partition_key_matches(const schema& base, const view_info& view, const dht::decorated_key& key, gc_clock::time_point now) {
```
The `now` parameter seems redundant now.  Maybe it can be reaped in another commit?

psarna replied:
Hm, now I wonder if it's correct that we just ignore `now`... I'd need to think if it's possible to get a dead row here (e.g. one which has a dead row marker). If it is, then this patch is incorrect, and if it isn't, then the `now` parameter is redundant indeed. I'll leave this verification for tomorrow.

psarna replied:
It looks ok - or at least consistent with previous code. Previously we indeed checked if cells are alive, but only for regular cells. Keys were assumed to be there, which also means that we assume the whole row is alive or otherwise it wouldn't even reach this layer. It also kind of makes sense, because why would anyone try to apply a filter on a dead row ¯\_(ツ)_/¯

avikivity replied:
`now` is used to convert atomic_cell to values, so if we do the conversion earlier, we don't need it here.

dekimir commented on db/view/view.cc:
```diff
@@ -259,15 +261,34 @@ void stats::register_stats() {
+            r->expression, exploded_pk, exploded_ck, dummy_row, &dummy_row, *selection, cql3::query_options({ }));
 }
 
 bool clustering_prefix_matches(const schema& base, const view_info& view, const partition_key& key, const clustering_key_prefix& ck, gc_clock::time_point now) {
```
Ditto.

psarna replied:
done

dekimir reviewed (commented):
Looks good, just a few minor comments.

psarna commented:
v2:
 - collapsed `get_value_from_partition_slice` to `get_value`
 - removed the now-redundant `now` parameter

psarna commented on cql3/expr/expression.cc:
```diff
@@ -168,19 +135,9 @@ managed_bytes_opt get_value_from_partition_slice(
-            std::bind(get_value_from_mutation, col, _1),
-            std::bind(get_value_from_partition_slice, col, _1, bag.options),
-        }, bag.row_data);
+    return get_value_from_partition_slice(col, bag.row_data, bag.options);
```
done

dekimir commented on db/view/view.cc:
```diff
@@ -318,6 +320,50 @@ static bool is_partition_key_empty(
+    void accept_new_partition(uint64_t row_count) {}
+
+    void accept_new_row(const clustering_key& key, const query::result_row_view& static_row, const query::result_row_view& row) {
+        _matches_view_filter = check_if_matches(static_row, row);
```
Is it correct to overwrite _matches_view_filter here, or to combine with the existing value?  If there are multiple calls to accept_new_row, we'll only see the last row's result.

dekimir replied:
Looking at how this visitor is used below, I believe accept_new_row will only be called once.  But this is fragile, and we can make it more robust by updating instead of overwriting _matches_view_filter, making this code more reusable tomorrow.

psarna replied:
Yes, the assumption is that we never check more than 1 row here. Still, updating instead of rewriting is trivial, so I'll do so

avikivity commented:
Wow, that was fast.

psarna commented:
> Wow, that was fast.

In contrast to the new path of generating views. But it's a good tradeoff.

avikivity commented on multishard_mutation_query.hh:
```diff
@@ -98,3 +100,24 @@ future<std::tuple<foreign_ptr<lw_shared_ptr<query::result>>, cache_temperature>>
         db::timeout_clock::time_point timeout);
+
+class data_query_result_builder {
+public:
```
It deserves its own header, can come later.

avikivity commented on db/view/view.cc:
```diff
@@ -318,6 +320,50 @@ static bool is_partition_key_empty(
+    void accept_new_partition(uint64_t row_count) {}
+
+    void accept_new_row(const clustering_key& key, const query::result_row_view& static_row, const query::result_row_view& row) {
+        _matches_view_filter &= check_if_matches(static_row, row);
```
@psarna you said you don't believe anyone uses view filtering. But isn't IS NOT NULL mandatory?

Perhaps it's implicit and/or stripped out. But logically, we can't create view rows when columns that become part of the view primary key are NULL.

psarna replied:
IS NOT NULL Is syntactic sugar (or rather syntactic bile™, since it uglifies the process of creating views), it doesn't evaluate to anything. But it **is** possible to create a filtering view like `create MATERIALIZED VIEW v AS SELECT * from t where v < 3 primary key (v,id);`, so we must support the filtering for backward compatibility if nothing else.

avikivity commented on db/view/view.cc:
```diff
@@ -318,6 +320,50 @@ static bool is_partition_key_empty(
+    void accept_partition_end(const query::result_row_view& static_row) {}
+
+    bool check_if_matches(const query::result_row_view& static_row, const query::result_row_view& row) const {
+        auto selection = cql3::selection::selection::wildcard(_base.shared_from_this());
```
Can this be moved to the constructor? It looks independent of static_row/row.

I'm guessing you re-create the visitor for each row, but perhaps that can be optimized.

avikivity replied:
It's probably negligible compared to all the other waste (exploding stuff, using bytes instead of managed_bytes_view), so we can defer it.

psarna replied:
It is negligible, but it's also a good idea, so I'll apply it - `wildcard()` actually creates a shared pointer, so it's better to make a useless allocation once, for future sake.

avikivity commented on db/view/view.cc:
```diff
@@ -318,6 +320,50 @@ static bool is_partition_key_empty(
+            _view.select_statement().get_restrictions()->get_non_pk_restriction() | boost::adaptors::map_values,
+            [&] (auto&& r) {
+                return cql3::expr::is_satisfied_by(
+                        r->expression, _pk, _ck, static_row, &row, *selection, cql3::query_options({ }));
```
Later, we can have just a single expression object representing all the filters.

dekimir replied:
Yes, that's inevitable since the type returned by get_non_pk_restrictions() will be deleted.  It will naturally fall to me to do it.

avikivity commented on db/view/view.cc:
```diff
@@ -332,14 +378,33 @@ static query::partition_slice make_partition_slice(const schema& s) {
+view_builder::build_step& view_builder::get_or_create_build_step(utils::UUID base_id) {
+    auto it = _base_to_build_step.find(base_id);
+    if (it == _base_to_build_step.end()) {
+        auto base = _db.find_column_family(base_id).shared_from_this();
```
This can potentially return a different base each time. Perhaps a base with different columns, or without the view. How safe are we against that?

psarna replied:
Github shows as if I created this function, but it's actually just code movement - whatever the issue might be here, it's preexisting. I think that view building gets restarted from scratch if the base table changes, but I'll need to verify it asynchronously to this series.

avikivity replied:
Sure.

avikivity commented on db/view/view.cc:
```diff
@@ -332,14 +378,33 @@ static query::partition_slice make_partition_slice(const schema& s) {
+
+    data_query_result_builder builder(base, slice, query::result_options::only_result(), query::result_memory_accounter{10*1024*1024});
+    builder.consume_new_partition(dht::decorate_key(base, key));
+    builder.consume(clustering_row(base, update), row_tombstone{}, update.is_live(base, tombstone{}, now));
```
Yes, so we recreate the builder each time. It's inefficient but nothing like the other inefficiencies here, so we can wait with improving it.

avikivity replied:
I see you don't pass the static row. Are views prohibited from filtering on the static row? Reasonable, but let's make sure.

psarna replied:
This function's signature didn't change and it does not pass any static row, so I think it's discarded earlier in the process. I'll try to verify.

psarna replied:
Verified in cqlsh. An attempt to restrict a static column in a view filter results in:
```
Non-primary key columns cannot be restricted in the SELECT statement used for materialized view v creation (got restrictions on: s)
```

avikivity commented on db/view/view.cc:
```diff
@@ -332,14 +378,33 @@ static query::partition_slice make_partition_slice(const schema& s) {
+    builder.consume(clustering_row(base, update), row_tombstone{}, update.is_live(base, tombstone{}, now));
+    builder.consume_end_of_partition();
+    auto result = builder.consume_end_of_stream();
+    view_filter_checking_visitor visitor(base, view,exploded_pk, exploded_ck);
```
In theory you could have folded these two visitors into one, but I totally understand why not.

avikivity commented:
`column_value_eval_bag` can now be dropped, but it's fine to leave it for @dekimir or me.

psarna commented:
> `column_value_eval_bag` can now be dropped, but it's fine to leave it for @dekimir or me.

I wouldn't dare deprive you two of this cherry on top of restrictions rewrite.

dekimir reviewed (approved):
LGTM

nyh commented on database.hh:
```diff
@@ -1001,7 +1001,7 @@ public:
     future<row_locker::lock_holder> do_push_view_replica_updates(schema_ptr s, mutation m, db::timeout_clock::time_point timeout, mutation_source source,
             tracing::trace_state_ptr tr_state, reader_concurrency_semaphore& sem, const io_priority_class& io_priority, query::partition_slice::option_set custom_opts) const;
-    std::vector<view_ptr> affected_views(const schema_ptr& base, const mutation& update, gc_clock::time_point now) const;
+    std::vector<view_ptr> affected_views(const schema_ptr& base, const mutation& update) const;
```
Why was "now" ever relevant to this function?

psarna replied:
It wasn't, it was only there because the previous interface of `is_satisfied_by` wanted it for regular cells, and this interface was used in `partition_key_matches`, which actually didn't provide any regular cells anyway. It would also have worked if we passed any time point, since it wasn't compared with anything.

nyh reviewed (approved):
I chose "approve" despite having to admit - I don't think I understood everything I saw, some of these APIs look so contrived :-( I hope our MV tests are good enough to find bugs in this relatively obscure area of filtered views. If not, we'll need to add more tests.

psarna commented on db/view/view.cc:
```diff
@@ -318,6 +320,50 @@ static bool is_partition_key_empty(
     }
 }
 
+// Checks if the result matches the provided view filter
```
Yes, it's worth noting that I only expect a single row here, will do.

psarna commented on db/view/view.cc:
```diff
@@ -318,6 +320,50 @@ static bool is_partition_key_empty(
+    }
+
+    void accept_new_row(const query::result_row_view& static_row, const query::result_row_view& row) {
+        _matches_view_filter &= check_if_matches(static_row, row);
```
You're right, I meant `&&=`, this bitwise operator just happens to work fine (the assumption is actually even lighter - that false is 0 and true is > 0), but I'll change to the proper one.

psarna commented on db/view/view.cc:
```diff
@@ -318,6 +320,50 @@ static bool is_partition_key_empty(
+        , _ck(ck)
+    {}
+
+    void accept_new_partition(const partition_key& key, uint64_t row_count) {}
```
The visitor interface is the most convenient way of accessing the underlying row data, but it's actually much more generic than what we need - including the fact that it also reads partition keys, etc - while we already know that we're going to run the visitor on a single row with well defined primary key. So, in short, this implementation is just a stub and the only sensible thing we could do here is to perhaps assert that the partition key is exactly the same as the one we passed in the constructor (which it is, because it comes from the same row). I chose not to assert here to avoid unnecessary panic checks in runtime, so I just left the function bodies empty.

And yes, I could omit passing the partition key in the constructor and instead read it here, but I prefer this visitor to be explicit about the fact that it currently only works for a single row and that its primary key is known at construction time.

psarna replied:
Hm, after hitting the "Comment" button I started having second thoughts about this interface. I'll give your suggestion a chance and use it in v3 if it works fine.

psarna replied:
It actually makes the visitor cleaner - thanks. I'll rerun all tests and push v3 with the suggestion.

psarna commented:
v3:
 - added a comment in the restriction checking visitor explaining that it expects a single row
 - instead of passing the keys in the constructor, the visitor pattern is used to obtain the keys
 - bitwise &= operator is replaced with a proper boolean one

psarna commented:
v4:
 - moved initializing CQL selection to the visitor constructor (doesn't change much since it's a visitor for a single row, but it's good practice)

nyh commented:
@psarna should I commit this v4, or are you still in a discussion with Avi on something?

psarna commented:
I see that I missed the request for moving the visitor to a separate header, which is a good idea. I'll push v5 in a few minutes.

psarna commented:
v5:
 - moved the data query visitor to a separate header

nyh commented:
Unfortunately after having merged this series, I had to dequeue now it because **debug mode** build failed in Jenkins - see https://jenkins.scylladb.com/job/scylla-master/job/next/3740/consoleText:
```
[2021-07-11T09:08:31.850Z] [800/3697] CXX build/debug/multishard_mutation_query.o
[2021-07-11T09:08:31.850Z] FAILED: build/debug/multishard_mutation_query.o 
[2021-07-11T09:08:31.850Z] clang++ -MD -MT build/debug/multishard_mutation_query.o -MF build/debug/multishard_mutation_query.o.d -I/jenkins/workspace/scylla-master/next/scylla/seastar/include -I/jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/gen/include -std=gnu++20 -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -DSEASTAR_API_LEVEL=6 -DSEASTAR_DEBUG -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_TYPE_ERASE_MORE -DFMT_LOCALE -DFMT_SHARED -I/usr/include/p11-kit-1  -DDEBUG -DSANITIZE -DDEBUG_LSA_SANITIZER -DSCYLLA_ENABLE_ERROR_INJECTION -Og -g -gz -iquote. -iquote build/debug/gen --std=gnu++20  -ffile-prefix-map=/jenkins/workspace/scylla-master/next/scylla=.  -march=westmere -DBOOST_TEST_DYN_LINK   -Iabseil -fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-parentheses-equality -Wno-c++11-narrowing -Wno-c++1z-extensions -Wno-sometimes-uninitialized -Wno-return-stack-address -Wno-missing-braces -Wno-unused-lambda-capture -Wno-misleading-indentation -Wno-overflow -Wno-noexcept-type -Wno-error=cpp -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-inconsistent-missing-override -Wno-defaulted-function-deleted -Wno-redeclared-class-member -Wno-unsupported-friend -Wno-unused-variable -Wno-delete-non-abstract-non-virtual-dtor -Wno-unknown-attributes -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-uninitialized-const-reference -Wno-psabi -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DHAVE_LZ4_COMPRESS_DEFAULT  -c -o build/debug/multishard_mutation_query.o multishard_mutation_query.cc
[2021-07-11T09:08:31.850Z] fatal error: error in backend: Cannot represent a difference across sections
[2021-07-11T09:08:31.850Z] PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
[2021-07-11T09:08:31.850Z] Stack dump:
[2021-07-11T09:08:31.850Z] 0.	Program arguments: /usr/bin/clang++ -std=gnu++20 -Werror=unused-result -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -Og -g -gz --std=gnu++20 -ffile-prefix-map=/jenkins/workspace/scylla-master/next/scylla=. -march=westmere -fvisibility=hidden -Wall -Wno-mismatched-tags -Wno-tautological-compare -Wno-parentheses-equality -Wno-c++11-narrowing -Wno-c++1z-extensions -Wno-sometimes-uninitialized -Wno-return-stack-address -Wno-missing-braces -Wno-unused-lambda-capture -Wno-misleading-indentation -Wno-overflow -Wno-noexcept-type -Wno-error=cpp -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-inconsistent-missing-override -Wno-defaulted-function-deleted -Wno-redeclared-class-member -Wno-unsupported-friend -Wno-unused-variable -Wno-delete-non-abstract-non-virtual-dtor -Wno-unknown-attributes -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-uninitialized-const-reference -Wno-psabi -Wno-error=deprecated-declarations -Werror -I/jenkins/workspace/scylla-master/next/scylla/seastar/include -I/jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/gen/include -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -DSEASTAR_API_LEVEL=6 -DSEASTAR_DEBUG -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_TYPE_ERASE_MORE -DFMT_LOCALE -DFMT_SHARED -I/usr/include/p11-kit-1 -DDEBUG -DSANITIZE -DDEBUG_LSA_SANITIZER -DSCYLLA_ENABLE_ERROR_INJECTION -iquote. -iquote build/debug/gen -DBOOST_TEST_DYN_LINK -Iabseil -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DHAVE_LZ4_COMPRESS_DEFAULT -c -o build/debug/multishard_mutation_query.o multishard_mutation_query.cc 
[2021-07-11T09:08:31.850Z] 1.	<eof> parser at end of file
[2021-07-11T09:08:31.850Z] 2.	Code generation
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamE+0x2a)[0x7f3b1d68727a]
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(_ZN4llvm3sys17RunSignalHandlersEv+0x34)[0x7f3b1d6854e4]
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(_ZN4llvm20CrashRecoveryContext10HandleExitEi+0x74)[0x7f3b1d5bbcc4]
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(_ZN4llvm3sys7Process4ExitEi+0x17)[0x7f3b1d6800b7]
[2021-07-11T09:08:31.850Z] /usr/bin/clang++(+0x144fe)[0x56293cca14fe]
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(_ZN4llvm18report_fatal_errorERKNS_5TwineEb+0x78)[0x7f3b1d5ca428]
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(+0x1d2fb63)[0x7f3b1e92bb63]
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(+0x1d10215)[0x7f3b1e90c215]
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(_ZN4llvm11MCAssembler11handleFixupERKNS_11MCAsmLayoutERNS_10MCFragmentERKNS_7MCFixupE+0xf1)[0x7f3b1e922e71]
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(_ZN4llvm11MCAssembler6layoutERNS_11MCAsmLayoutE+0x2a1)[0x7f3b1e923fc1]
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(_ZN4llvm11MCAssembler6FinishEv+0x39)[0x7f3b1e9241a9]
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(_ZN4llvm10AsmPrinter14doFinalizationERNS_6ModuleE+0x7c1)[0x7f3b1dde6261]
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(_ZN4llvm13FPPassManager14doFinalizationERNS_6ModuleE+0x5d)[0x7f3b1d7914fd]
[2021-07-11T09:08:31.850Z] /lib64/libLLVM-11.so(_ZN4llvm6legacy15PassManagerImpl3runERNS_6ModuleE+0x460)[0x7f3b1d79c680]
[2021-07-11T09:08:31.851Z] /lib64/libclang-cpp.so.11(_ZN5clang17EmitBackendOutputERNS_17DiagnosticsEngineERKNS_19HeaderSearchOptionsERKNS_14CodeGenOptionsERKNS_13TargetOptionsERKNS_11LangOptionsERKN4llvm10DataLayoutEPNSE_6ModuleENS_13BackendActionESt10unique_ptrINSE_17raw_pwrite_streamESt14default_deleteISM_EE+0x14df)[0x7f3b2332a72f]
[2021-07-11T09:08:31.851Z] /lib64/libclang-cpp.so.11(+0x16c0762)[0x7f3b235b4762]
[2021-07-11T09:08:31.851Z] /lib64/libclang-cpp.so.11(_ZN5clang8ParseASTERNS_4SemaEbb+0x29d)[0x7f3b227dfb2d]
[2021-07-11T09:08:31.851Z] /lib64/libclang-cpp.so.11(_ZN5clang14FrontendAction7ExecuteEv+0xb1)[0x7f3b23b94471]
[2021-07-11T09:08:31.851Z] /lib64/libclang-cpp.so.11(_ZN5clang16CompilerInstance13ExecuteActionERNS_14FrontendActionE+0x1e3)[0x7f3b23b679d3]
[2021-07-11T09:08:31.851Z] /lib64/libclang-cpp.so.11(_ZN5clang25ExecuteCompilerInvocationEPNS_16CompilerInstanceE+0x6d0)[0x7f3b23bf6010]
[2021-07-11T09:08:31.851Z] /usr/bin/clang++(_Z8cc1_mainN4llvm8ArrayRefIPKcEES2_Pv+0x910)[0x56293cca1e80]
[2021-07-11T09:08:31.851Z] /usr/bin/clang++(+0x17b09)[0x56293cca4b09]
[2021-07-11T09:08:31.851Z] /lib64/libclang-cpp.so.11(+0x19b5415)[0x7f3b238a9415]
[2021-07-11T09:08:31.851Z] /lib64/libLLVM-11.so(_ZN4llvm20CrashRecoveryContext9RunSafelyENS_12function_refIFvvEEE+0x23)[0x7f3b1d5bbb53]
[2021-07-11T09:08:31.851Z] /lib64/libclang-cpp.so.11(_ZNK5clang6driver10CC1Command7ExecuteEN4llvm8ArrayRefINS2_8OptionalINS2_9StringRefEEEEEPNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPb+0x14f)[0x7f3b238bc9ef]
[2021-07-11T09:08:31.851Z] /lib64/libclang-cpp.so.11(_ZNK5clang6driver11Compilation14ExecuteCommandERKNS0_7CommandERPS3_+0x97)[0x7f3b238c2e07]
[2021-07-11T09:08:31.851Z] /lib64/libclang-cpp.so.11(_ZNK5clang6driver11Compilation11ExecuteJobsERKNS0_7JobListERN4llvm15SmallVectorImplISt4pairIiPKNS0_7CommandEEEE+0x73)[0x7f3b238c31c3]
[2021-07-11T09:08:31.851Z] /lib64/libclang-cpp.so.11(_ZN5clang6driver6Driver18ExecuteCompilationERNS0_11CompilationERN4llvm15SmallVectorImplISt4pairIiPKNS0_7CommandEEEE+0xe3)[0x7f3b238c4503]
[2021-07-11T09:08:31.851Z] /usr/bin/clang++(main+0x2277)[0x56293cc9e497]
[2021-07-11T09:08:31.851Z] /lib64/libc.so.6(__libc_start_main+0xf2)[0x7f3b1c8561e2]
[2021-07-11T09:08:31.851Z] /usr/bin/clang++(_start+0x2e)[0x56293cc9f55e]
```
This seems to be a clang bug which is specific to the version of clang running on the Jenkins machine. On my laptop with Fedora 34 and clang 12.0 I see a similar error message but it is apparently (?) not fatal:
```
$ /usr/bin/clang++ -MD -MT build/debug/multishard_mutation_query.o -MF build/debug/multishard_mutation_query.o.d -Iseastar/include -Ibuild/debug/seastar/gen/include -std=gnu++20 -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -DSEASTAR_API_LEVEL=6 -DSEASTAR_DEBUG -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_TYPE_ERASE_MORE -DFMT_LOCALE -DFMT_SHARED -I/usr/include/p11-kit-1  -DDEBUG -DSANITIZE -DDEBUG_LSA_SANITIZER -DSCYLLA_ENABLE_ERROR_INJECTION -Og -g -gz -iquote. -iquote build/debug/gen --std=gnu++20  -ffile-prefix-map=/jenkins/workspace/scylla-master/next/scylla=.  -march=westmere -DBOOST_TEST_DYN_LINK   -Iabseil -fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-parentheses-equality -Wno-c++11-narrowing -Wno-c++1z-extensions -Wno-sometimes-uninitialized -Wno-return-stack-address -Wno-missing-braces -Wno-unused-lambda-capture -Wno-misleading-indentation -Wno-overflow -Wno-noexcept-type -Wno-error=cpp -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-inconsistent-missing-override -Wno-defaulted-function-deleted -Wno-redeclared-class-member -Wno-unsupported-friend -Wno-unused-variable -Wno-delete-non-abstract-non-virtual-dtor -Wno-unknown-attributes -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-uninitialized-const-reference -Wno-psabi -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DHAVE_LZ4_COMPRESS_DEFAULT  -c -o build/debug/multishard_mutation_query.o multishard_mutation_query.cc
<unknown>:0: error: Cannot represent a difference across sections
<unknown>:0: error: Cannot represent a difference across sections
<unknown>:0: error: Cannot represent a difference across sections
```

I don't know what piece of the code exercise this clang bug, but maybe with some slight modification or ifdef (the compilation only fails in debug mode) we can avoid it.

avikivity commented:
> 
> This seems to be a clang bug which is specific to the version of clang running on the Jenkins machine. On my laptop with Fedora 34 and clang 12.0 I see a similar error message but it is apparently (?) not fatal:

We don't have a specific version running on Jenkins. Jenkins instead uses the frozen toolchain that is recorded in scylla.git.

avikivity commented:
I'd like to update to Fedora 34/clang 12 but for that two things need to happen

1. Fedora updates gcc 11 to include the fix for https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100900
2. Fedora updates clang to 12.0.1 to include the fix for https://bugs.llvm.org/show_bug.cgi?id=48857

We really only need to wait for 1, and regenerate the toolchain again when 2 happens.

avikivity commented:
I filed https://bugzilla.redhat.com/show_bug.cgi?id=1981085 to kick off 1.

psarna commented:
@avikivity @nyh for whatever reason (inlining would be my guess), the issue disappears if I blatantly copy-paste the `data_query_result_builder` class from multishard_mutation_query.cc straight to view.cc. While copy-pasting is bad, this visitor is also a really simple wrapper, so perhaps it's better than not being able to proceed with this series due to a compiler bug? Opinions welcome, I can push v6 for preview.

avikivity commented:
Sure, let's make progress.

nyh commented:
> @avikivity @nyh for whatever reason (inlining would be my guess), the issue disappears if I blatantly copy-paste the `data_query_result_builder` class from multishard_mutation_query.cc straight to view.cc.

This is copying instead of what? Instead of #including a common header file? Or what?

I'm not against copying it. Especially if you can remove pieces of it you don't actually need and the result is even shorter.

nyh commented:
By the way @psarna if you still have the broken patched file hanging around, and can preprocess it to one big preprocessed source file (by adding -E to the compilation command), you can report it as a clang bug.

psarna commented:
v6:
 - instead of moving the query builder class to a public header, a simplified version of it is used directly in view.cc

@nyh Re reporting a clang bug: I still have that branch, but first I'll need some time to verify if it's not a duplicate for something that was already reported by us (e.g. https://bugs.llvm.org/show_bug.cgi?id=50345 shows a similar inability to represent a difference across sections)

nyh commented on db/view/view.cc:
```diff
@@ -331,6 +331,32 @@ static query::partition_slice make_partition_slice(const schema& s) {
+class data_query_result_builder {
+public:
+    using result_type = query::result;
+    static constexpr emit_only_live_rows only_live = emit_only_live_rows::yes;
```
Do you actually use this variable? I don't see where.

psarna replied:
I didn't dive into details, but I assumed that both this static and the result_type typedef above might be needed by the visitor pattern for some reason - e.g. I imagine a situation where the visitor type is a template parameter for the visiting function, and this function requires that these two are present. I didn't verify that assumption though and simply left these two intact just in case, perhaps they can be removed/inlined.

nyh commented on db/view/view.cc:
```diff
@@ -331,6 +331,32 @@ static query::partition_slice make_partition_slice(const schema& s) {
 
+class data_query_result_builder {
+public:
+    using result_type = query::result;
```
nitpick - only used once... So you just do the replacement manually.

nyh replied:
Or, maybe you will use it with some template that needs these definitions?

psarna commented:
I'll self-dequeue, the series hasn't passed debug unit tests (which I should have run locally first). I'm also leaving a note for myself for the debugging session tomorrow:
```
READ of size 8 at 0x7f9285937b20 thread T1
    #0 0x11b10a0e in seastar::lw_shared_ptr<cql3::column_specification>::lw_shared_ptr(seastar::lw_shared_ptr<cql3::column_specification> const&) /tmp/./seastar/include/seastar/core/shared_ptr.hh:290:59
    #1 0x133bef31 in decltype(new ((void*)(0))seastar::lw_shared_ptr<cql3::column_specification>(std::declval<seastar::lw_shared_ptr<cql3::column_specification> const&>())) std::construct_at<seastar::lw_shared_ptr<cql3::column_specification>, seastar::lw_shared_ptr<cql3::column_specification> const&
>(seastar::lw_shared_ptr<cql3::column_specification>*, seastar::lw_shared_ptr<cql3::column_specification> const&) /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/stl_construct.h:97:39
    #2 0x133be54e in void std::allocator_traits<std::allocator<seastar::lw_shared_ptr<cql3::column_specification> > >::construct<seastar::lw_shared_ptr<cql3::column_specification>, seastar::lw_shared_ptr<cql3::column_specification> const&>(std::allocator<seastar::lw_shared_ptr<cql3::column_specifica
tion> >&, seastar::lw_shared_ptr<cql3::column_specification>*, seastar::lw_shared_ptr<cql3::column_specification> const&) /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/alloc_traits.h:514:4
    #3 0x1339b609 in seastar::lw_shared_ptr<cql3::column_specification>& std::vector<seastar::lw_shared_ptr<cql3::column_specification>, std::allocator<seastar::lw_shared_ptr<cql3::column_specification> > >::emplace_back<seastar::lw_shared_ptr<cql3::column_specification> const&>(seastar::lw_shared_p
tr<cql3::column_specification> const&) /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/vector.tcc:115:6
    #4 0x13f31d4d in cql3::selection::simple_selection::make(seastar::lw_shared_ptr<schema const>, std::vector<column_definition const*, std::allocator<column_definition const*> >, bool) /tmp/cql3/selection/selection.cc:99:22
    #5 0x13f25d59 in cql3::selection::selection::for_columns(seastar::lw_shared_ptr<schema const>, std::vector<column_definition const*, std::allocator<column_definition const*> >) /tmp/cql3/selection/selection.cc:240:12
    #6 0x148b938c in db::view::partition_key_matches(schema const&, view_info const&, dht::decorated_key const&) /tmp/db/view/view.cc:269:22
    #7 0x118b040c in auto table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53::operator()<view_ptr const&>(view_ptr const&) const table.cc:1639:16
    #8 0x118b025d in bool boost::range_detail::default_constructible_unary_fn_wrapper<table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53, bool>::operator()<view_ptr>(view_ptr const&) const /usr/include/boost/range/detail/default_constructible_unary_fn.hpp
:56:16
    #9 0x118b011a in boost::iterators::filter_iterator<boost::range_detail::default_constructible_unary_fn_wrapper<table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53, bool>, __gnu_cxx::__normal_iterator<view_ptr const*, std::vector<view_ptr, std::allocato
r<view_ptr> > > >::satisfy_predicate() /usr/include/boost/iterator/filter_iterator.hpp:100:50
    #10 0x118afc35 in boost::iterators::filter_iterator<boost::range_detail::default_constructible_unary_fn_wrapper<table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53, bool>, __gnu_cxx::__normal_iterator<view_ptr const*, std::vector<view_ptr, std::allocator<view_ptr> > > >::filter_iterator(boost::range_detail::default_constructible_unary_fn_wrapper<table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53, bool>, __gnu_cxx::__normal_iterator<view_ptr const*, std::vector<view_ptr, std::allocator<view_ptr> > >, __gnu_cxx::__normal_iterator<view_ptr const*, std::vector<view_ptr, std::allocator<view_ptr> > >) /usr/include/boost/iterator/filter_iterator.hpp:59:11
    #11 0x118af6e2 in boost::iterators::filter_iterator<boost::range_detail::default_constructible_unary_fn_wrapper<table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53, bool>, __gnu_cxx::__normal_iterator<view_ptr const*, std::vector<view_ptr, std::allocator<view_ptr> > > > boost::iterators::make_filter_iterator<boost::range_detail::default_constructible_unary_fn_wrapper<table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53, bool>, __gnu_cxx::__normal_iterator<view_ptr const*, std::vector<view_ptr, std::allocator<view_ptr> > > >(boost::range_detail::default_constructible_unary_fn_wrapper<table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53, bool>, __gnu_cxx::__normal_iterator<view_ptr const*, std::vector<view_ptr, std::allocator<view_ptr> > >, __gnu_cxx::__normal_iterator<view_ptr const*, std::vector<view_ptr, std::allocator<view_ptr> > >) /usr/include/boost/iterator/filter_iterator.hpp:114:14
    #12 0x118af42b in boost::range_detail::filtered_range<table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53, std::vector<view_ptr, std::allocator<view_ptr> > const>::filtered_range(table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53, std::vector<view_ptr, std::allocator<view_ptr> > const&) /usr/include/boost/range/adaptor/filtered.hpp:45:20
    #13 0x1181bf80 in boost::range_detail::filtered_range<table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53, std::vector<view_ptr, std::allocator<view_ptr> > const> boost::range_detail::operator|<std::vector<view_ptr, std::allocator<view_ptr> >, table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53>(std::vector<view_ptr, std::allocator<view_ptr> > const&, boost::range_detail::filter_holder<table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const::$_53> const&) /usr/include/boost/range/adaptor/filtered.hpp:75:20
    #14 0x1181ba7c in table::affected_views(seastar::lw_shared_ptr<schema const> const&, mutation const&) const table.cc:1638:60
    #15 0x11828b53 in table::do_push_view_replica_updates(seastar::lw_shared_ptr<schema const>, mutation, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >, mutation_source, tracing::trace_state_ptr, reader_concurrency_semaphore&, seastar::io_priority_class const&, enum_set<super_enum<query::partition_slice::option, (query::partition_slice::option)0, (query::partition_slice::option)1, (query::partition_slice::option)2, (query::partition_slice::option)3, (query::partition_slice::option)4, (query::partition_slice::option)5, (query::partition_slice::option)6, (query::partition_slice::option)7, (query::partition_slice::option)8, (query::partition_slice::option)9, (query::partition_slice::option)10, (query::partition_slice::option)11, (query::partition_slice::option)12> >) const table.cc:2245:52
    #16 0x1182820b in table::push_view_replica_updates(seastar::lw_shared_ptr<schema const> const&, mutation&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >, tracing::trace_state_ptr, reader_concurrency_semaphore&) const table.cc:2290:12
    #17 0x11827df3 in table::push_view_replica_updates(seastar::lw_shared_ptr<schema const> const&, frozen_mutation const&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >, tracing::trace_state_ptr, reader_concurrency_semaphore&) const table.cc:2225:12
    #18 0x11534d59 in database::do_apply(seastar::lw_shared_ptr<schema const>, frozen_mutation const&, tracing::trace_state_ptr, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >, seastar::bool_class<force_sync_tag>) /tmp/database.cc:1829:44
    #19 0x1172763f in seastar::future<void> std::__invoke_impl<seastar::future<void>, seastar::future<void> (database::* const&)(seastar::lw_shared_ptr<schema const>, frozen_mutation const&, tracing::trace_state_ptr, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >, seastar::bool_class<force_sync_tag>), database*, seastar::lw_shared_ptr<schema const>, frozen_mutation const&, tracing::trace_state_ptr, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >, seastar::bool_class<force_sync_tag> >(std::__invoke_memfun_deref, seastar::future<void> (database::* const&)(seastar::lw_shared_ptr<schema const>, frozen_mutation const&, tracing::trace_state_ptr, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >, seastar::bool_class<force_sync_tag>), database*&&, seastar::lw_shared_ptr<schema const>&&, frozen_mutation const&, tracing::trace_state_ptr&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >&&, seastar::bool_class<force_sync_tag>&&) /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:73:14

```

psarna commented:
v7:
 - fixed the debug tests failure: due to the use of `auto` instead of `const auto&` in the foor loop, the code sometimes compiled to taking the address of a temporary copy and then relied on dereferencing it; now fixed

Note: my local debug tests are still running because it takes eternity for them to finish, but a sample of previously failing tests succeeded, so I'm publishing this version

psarna commented:
ok, green light, the debug tests passed

nyh commented:
Pushed again. Good luck!
